src/Grpc.Net.Client/Internal/Retry/RetryCallBase.cs

```diff
@@ -0,0 +1,518 @@
+using Grpc.Net.Client.Internal.Http;
+using Grpc.Shared;
+using Microsoft.Extensions.Logging;
+
+#if NETSTANDARD2_0
+using ValueTask = System.Threading.Tasks.Task;
+#endif
+
+namespace Grpc.Net.Client.Internal.Retry
+{
+    internal abstract partial class RetryCallBase<TRequest, TResponse> : IGrpcCall<TRequest, TResponse>
+        where TRequest : class
+        where TResponse : class
+    {
+        private readonly TaskCompletionSource<IGrpcCall<TRequest, TResponse>> _commitedCallTcs;
+        private RetryCallBaseClientStreamReader<TRequest, TResponse>? _retryBaseClientStreamReader;
+        private RetryCallBaseClientStreamWriter<TRequest, TResponse>? _retryBaseClientStreamWriter;
+        private CancellationTokenRegistration? _ctsRegistration;
+
+        protected object Lock { get; } = new object();
+        protected ILogger Logger { get; }
+        protected Method<TRequest, TResponse> Method { get; }
+        protected CallOptions Options { get; }
+        protected int MaxRetryAttempts { get; }
+        protected CancellationTokenSource CancellationTokenSource { get; }
+        protected TaskCompletionSource<IGrpcCall<TRequest, TResponse>?>? NewActiveCallTcs { get; set; }
+        protected bool Disposed { get; private set; }
+
+        public GrpcChannel Channel { get; }
+        public Task<IGrpcCall<TRequest, TResponse>> CommitedCallTask => _commitedCallTcs.Task;
+        public IAsyncStreamReader<TResponse>? ClientStreamReader => _retryBaseClientStreamReader ??= new RetryCallBaseClientStreamReader<TRequest, TResponse>(this);
+        public IClientStreamWriter<TRequest>? ClientStreamWriter => _retryBaseClientStreamWriter ??= new RetryCallBaseClientStreamWriter<TRequest, TResponse>(this);
+        public WriteOptions? ClientStreamWriteOptions { get; internal set; }
+        public bool ClientStreamComplete { get; set; }
+
+        protected int AttemptCount { get; private set; }
+        protected List<ReadOnlyMemory<byte>> BufferedMessages { get; }
+        protected long CurrentCallBufferSize { get; set; }
+        protected bool BufferedCurrentMessage { get; set; }
+
+        protected RetryCallBase(GrpcChannel channel, Method<TRequest, TResponse> method, CallOptions options, string loggerName, int retryAttempts)
+        {
+            Logger = channel.LoggerFactory.CreateLogger(loggerName);
+            Channel = channel;
+            Method = method;
+            Options = options;
+            _commitedCallTcs = new TaskCompletionSource<IGrpcCall<TRequest, TResponse>>(TaskCreationOptions.RunContinuationsAsynchronously);
+            BufferedMessages = new List<ReadOnlyMemory<byte>>();
+
+            // If the passed in token is canceled then we want to cancel the retry cancellation token.
+            if (options.CancellationToken.CanBeCanceled)
+            {
+                _ctsRegistration = options.CancellationToken.Register(state => ((RetryCallBase<TRequest, TResponse>)state!).CancellationTokenSource.Cancel(), this);
```

Comment:
This looks like this will nullref if `options.CancellationToken` is already cancelled causing the Register callback runs inline.

It's probably best to just construct the CTS and register in one go when `options.CancellationToken.CanBeCanceled` with `CreateLinkedTokenSource`. Unless you want to avoid disposing the CTS which I understand if that's the case.